core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -884,16 +881,27 @@ private void internalPerformTranslogRecovery(boolean skipTranslogRecovery, boole
             }
         }
         recoveryState.setStage(RecoveryState.Stage.TRANSLOG);
+        final EngineConfig.OpenMode openMode;
+        if (indexExists == false) {
+            openMode = EngineConfig.OpenMode.CREATE_INDEX_AND_TRANSLOG;
+        } else if (skipTranslogRecovery) {
+            openMode = EngineConfig.OpenMode.OPEN_INDEX_CREATE_TRANSLOG;
+        } else {
+            openMode = EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG;
+        }
+        final EngineConfig config = newEngineConfig(openMode, translogConfig, cachingPolicy,
+            new IndexShardRecoveryPerformer(shardId, mapperService, logger));
         // we disable deletes since we allow for operations to be executed against the shard while recovering
         // but we need to make sure we don't loose deletes until we are done recovering
-        engineConfig.setEnableGcDeletes(false);
-        engineConfig.setCreate(indexExists == false);
-        if (skipTranslogRecovery == false) {
+        config.setEnableGcDeletes(false);
+        Engine newEngine = createNewEngine(config);
+        verifyNotClosed();
+        if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
```

Comment:
I'm talking about the the creation of a new shard, where we "recover" from store (maybe we should change that, it's confusing - different change) and we end up here where both indexExists and skipTranslogRecovery are false. In this case we use to create and engine, create a trasnlog, recover from it (does nothing because it's empty) and commit the translog into the empty engine. Now we don't do the last commit. We don't do it here (because openMode == CREATE_INDEX_AND_TRANSLOG). The engine doesn't do it because:

```
 if (openMode == EngineConfig.OpenMode.OPEN_INDEX_CREATE_TRANSLOG) {
                    // make sure we point at the latest translog from now on..
                    commitIndexWriter(writer, translog, lastCommittedSegmentInfos.getUserData().get(SYNC_COMMIT_ID));
                }
```